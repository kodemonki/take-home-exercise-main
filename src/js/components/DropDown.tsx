import React, { useEffect, useState } from "react";
import config from "../config";
import { Transform } from "supertest/lib/test";

const container = {
  display: "inline-block",
  position: "relative" as const,
};

const button = {
  background: "WhiteSmoke",
  borderBottom: "2px solid LightGrey",
  padding: "10px 20px",
  cursor: "pointer",
};

const content = {
  background: "WhiteSmoke",
  position: "absolute" as const,
  maxHeight: "450px",
  overflowY: "scroll" as const,
};

const rowItem = {
  padding: "10px 20px",
  minWidth: "120px",
};

const stretchedChar = {
  transform: "scaleX(2) scaleY(-1) translate(4px,2px)" as const,
  display: "inline-block",
};

interface DDProps {
  title: string;
  data: string[];
}

const DropDown: React.FC<DDProps> = ({ title, data }) => {
  const [panelVisibility, setPanelVisibility] = useState(
    ("hidden" as const) || ("visible" as const)
  );
  const handleClick = () => {
    if (panelVisibility === "hidden") {
      setPanelVisibility("visible");
    } else {
      setPanelVisibility("hidden");
    }
  };

  return (
    <div style={container}>
      <div onClick={handleClick} style={button}>
        <b>{title}</b> <span style={stretchedChar}>^</span>
      </div>
      <div style={{ ...content, ...{ visibility: panelVisibility } }}>
        <div style={rowItem}>
          <input
            type="checkbox"
            id={`${title}all`}
            name={`${title}all`}
            value="all"
          />
          <label htmlFor={`${title}all`}>all</label>
        </div>
        {data.map((item, index) => {
          return (
            <div key={item} style={rowItem}>
              <input
                type="checkbox"
                id={`${title}${index}`}
                name={`${title}${index}`}
                value={item}
              />
              <label htmlFor={`${title}${index}`}>{item}</label>
              <br />
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default DropDown;
