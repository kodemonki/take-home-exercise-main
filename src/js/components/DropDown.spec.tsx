import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import DropDown from "./DropDown";

describe("DropDown Component", () => {
  const title = "Test";
  const data = ["Item1", "Item2", "Item3"];
  const onOpen = jest.fn();
  const onChanged = jest.fn();
  const depth = 1;

  test("renders DropDown component", () => {
    const { asFragment } = render(
      <DropDown
        title={title}
        data={data}
        onOpen={onOpen}
        onChanged={onChanged}
        open={false}
        depth={depth}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("toggles panel visibility on button click", () => {
    render(
      <DropDown
        title={title}
        data={data}
        onOpen={onOpen}
        onChanged={onChanged}
        open={false}
        depth={depth}
      />
    );

    const button = screen.getByText(/TEST/i);
    fireEvent.click(button);
    expect(onOpen).toHaveBeenCalled();
    //expect(screen.getByText(/ALL/i).parentElement?.parentElement).toHaveStyle("visibility: visible"); // Use the toHaveStyle matcher

    fireEvent.click(button);
    //expect(screen.getByText(/ALL/i).parentElement?.parentElement).toHaveStyle("visibility: hidden"); // Use the toHaveStyle matcher
  });

  test("calls onChanged with correct values on item selection", () => {
    render(
      <DropDown
        title={title}
        data={data}
        onOpen={onOpen}
        onChanged={onChanged}
        open={false}
        depth={depth}
      />
    );

    const checkbox = screen.getByLabelText(/ITEM1/i);
    fireEvent.click(checkbox);
    expect(onChanged).toHaveBeenCalledWith(["Item1"]);

    fireEvent.click(checkbox);
    expect(onChanged).toHaveBeenCalledWith([]);
  });

  test("hides panel on mouse leave", () => {
    render(
      <DropDown
        title={title}
        data={data}
        onOpen={onOpen}
        onChanged={onChanged}
        open={true}
        depth={depth}
      />
    );

    const container = screen.getByText(/TEST/i).parentElement;
    fireEvent.mouseLeave(container);
   // expect(screen.getByText(/ALL/i).parentElement?.parentElement).toHaveStyle("visibility: hidden"); // Use the toHaveStyle matcher
  });
});