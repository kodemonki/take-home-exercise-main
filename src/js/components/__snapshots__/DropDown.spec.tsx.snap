// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown Component renders DropDown component 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; position: relative; z-index: 1;"
  >
    <div
      style="background: whiteSmoke; border-bottom: 2px solid LightGrey; padding: 10px 20px; cursor: pointer;"
    >
       
      <b>
        TEST
      </b>
       
      <span
        style="transform: scaleX(2) scaleY(-1) translate(4px,2px); display: inline-block;"
      >
        ^
      </span>
    </div>
    <div
      style="box-sizing: border-box; background: whiteSmoke; position: absolute; max-height: 450px; overflow-y: scroll; top: 60px; border: 1px solid lightgrey; visibility: hidden;"
    >
      <div
        style="padding: 10px 20px; min-width: 80px; border-bottom: 1px solid GainsBoro; cursor: pointer;"
      >
        <label
          class="checkboxContainer"
        >
          <input
            id="Testall"
            name="Testall"
            type="checkbox"
            value="all"
          />
          <span
            class="checkmark"
          />
          <b>
            ALL
          </b>
        </label>
      </div>
      <div
        style="padding: 10px 20px; min-width: 80px; border-bottom: 1px solid GainsBoro; cursor: pointer;"
      >
        <label
          class="checkboxContainer"
        >
          <input
            id="Test0"
            name="Test0"
            type="checkbox"
            value="Item1"
          />
          <span
            class="checkmark"
          />
          <b>
            ITEM1
          </b>
        </label>
      </div>
      <div
        style="padding: 10px 20px; min-width: 80px; border-bottom: 1px solid GainsBoro; cursor: pointer;"
      >
        <label
          class="checkboxContainer"
        >
          <input
            id="Test1"
            name="Test1"
            type="checkbox"
            value="Item2"
          />
          <span
            class="checkmark"
          />
          <b>
            ITEM2
          </b>
        </label>
      </div>
      <div
        style="padding: 10px 20px; min-width: 80px; border-bottom: 1px solid GainsBoro; cursor: pointer;"
      >
        <label
          class="checkboxContainer"
        >
          <input
            id="Test2"
            name="Test2"
            type="checkbox"
            value="Item3"
          />
          <span
            class="checkmark"
          />
          <b>
            ITEM3
          </b>
        </label>
      </div>
    </div>
    <div
      style="overflow: hidden; display: block; width: 60px; height: 20px; visibility: hidden;"
    >
      <div
        style="width: 50px; height: 50px; display: block; background: whiteSmoke; transform: rotate(45deg); border: 1px solid gainsboro; margin-top: 15px; margin-left: 6px; visibility: hidden;"
      />
    </div>
  </div>
</DocumentFragment>
`;
